kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/Polynomial.kt

```diff
@@ -5,128 +5,484 @@
 
 package space.kscience.kmath.functions
 
-import space.kscience.kmath.misc.UnstableKMathAPI
 import space.kscience.kmath.operations.*
-import kotlin.contracts.InvocationKind
-import kotlin.contracts.contract
 import kotlin.math.max
-import kotlin.math.pow
+import kotlin.math.min
 
 /**
- * Polynomial coefficients model without fixation on specific context they are applied to.
+ * Polynomial model without fixation on specific context they are applied to.
  *
  * @param coefficients constant is the leftmost coefficient.
  */
-public class Polynomial<out T>(public val coefficients: List<T>) {
+public data class Polynomial<C>(
+    /**
+     * List that collects coefficients of the polynomial. Every monomial `a x^d` is represented as a coefficients
+     * `a` placed into the list with index `d`. For example coefficients of polynomial `5 x^2 - 6` can be represented as
+     * ```
+     * listOf(
+     *     -6, // -6 +
+     *     0,  // 0 x +
+     *     5,  // 5 x^2
+     * )
+     * ```
+     * and also as
+     * ```
+     * listOf(
+     *     -6, // -6 +
+     *     0,  // 0 x +
+     *     5,  // 5 x^2
+     *     0,  // 0 x^3
+     *     0,  // 0 x^4
+     * )
+     * ```
+     * It is recommended not to put extra zeros at end of the list (as for `0x^3` and `0x^4` in the example), but is not
+     * prohibited.
+     */
+    public val coefficients: List<C>
+) : AbstractPolynomial<C> {
     override fun toString(): String = "Polynomial$coefficients"
 }
 
-/**
- * Returns a [Polynomial] instance with given [coefficients].
- */
-@Suppress("FunctionName")
-public fun <T> Polynomial(vararg coefficients: T): Polynomial<T> = Polynomial(coefficients.toList())
+// region Internal utilities
 
 /**
- * Evaluates the value of the given double polynomial for given double argument.
+ * Represents internal [Polynomial] errors.
  */
-public fun Polynomial<Double>.value(arg: Double): Double = coefficients.reduceIndexed { index, acc, c ->
-    acc + c * arg.pow(index)
+internal class PolynomialError : Error {
```

Comment:
Thank you for the comment. But there are 2 moments:
1. Actually, I did not use it, so I decided just to remove it.
2. There is no documentation for `kotlin.Error`. Only in K/JVM implementation it is a type alias for `java.lang.Error` that is described as (I look in [Java 11 API](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Error.html)):
    > An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it.
    
    And I was going to use it in cases of strange internal behaviour of polynomials (like in case when you have more complicated structure inside `Polynomial`, and it was used/changed incorrectly, so it looks like addition of two polynomials suddenly crashes). Such cases meet the definition, so `Error` is right choice. Or do I miss something?

Reply:
> , and it was used/changed incorrectly

The convention is that `Error` is used only for abnormal situations related to VM (incorrect bytecode, out-of-memory). Passing incorrect arguments to a function is clearly a `RuntimeException`.

Reply:
> The convention is that `Error` is used only for abnormal situations related to VM (incorrect bytecode, out-of-memory).

OK. Got it.

> Passing incorrect arguments to a function is clearly a `RuntimeException`.

I was talking not about case of passing incorrect arguments, but about case when some "system" made mistake that it supposedly should not make.